src/Microsoft.Diagnostics.Monitoring.WebApi/CollectionRulePipelineState.cs

```diff
@@ -0,0 +1,231 @@
+                    CurrentStateReason = Strings.Message_CollectionRuleStateReason_Running;
+                }
+            }
+        }
+
+        public void StartupTriggerCompleted()
+        {
+            lock (_lock)
+            {
+                Debug.Assert(CurrentState != CollectionRuleState.Finished);
+
+                CurrentState = CollectionRuleState.Finished;
+                CurrentStateReason = Strings.Message_CollectionRuleStateReason_Finished_Startup;
+            }
+        }
+
+        public void RuleDurationReached()
+        {
+            lock (_lock)
+            {
+                Debug.Assert(CurrentState != CollectionRuleState.Finished);
+
+                CurrentState = CollectionRuleState.Finished;
+                CurrentStateReason = Strings.Message_CollectionRuleStateReason_Finished_RuleDuration;
+            }
+        }
+
+        private void ActionCountReached()
+        {
+            lock (_lock)
+            {
+                CurrentState = CollectionRuleState.Finished;
+                CurrentStateReason = Strings.Message_CollectionRuleStateReason_Finished_ActionCount;
+            }
+        }
+
+        public void AddTimestamp(DateTime currentTimestamp)
+        {
+            lock (_lock)
+            {
+                ExecutionTimestamps.Enqueue(currentTimestamp);
+                AllExecutionTimestamps.Add(currentTimestamp);
+            }
+        }
+
+        public void RuleFailure(string errorMessage)
+        {
+            lock (_lock)
+            {
+                Debug.Assert(CurrentState != CollectionRuleState.Finished);
+
+                CurrentState = CollectionRuleState.Finished;
+                CurrentStateReason = string.Format(CultureInfo.InvariantCulture, Strings.Message_CollectionRuleStateReason_Finished_Failure, errorMessage);
+            }
+        }
+
+        public bool CanExecuteActions(DateTime currentTime)
+        {
+            lock (_lock)
+            {
+                DequeueOldTimestamps(ExecutionTimestamps, ActionCountSlidingWindowDuration, currentTime);
+            }
+            bool canExecuteActions = !CheckForThrottling(ActionCountLimit, ActionCountSlidingWindowDuration, ExecutionTimestamps.Count);
+
+            if (canExecuteActions)
+            {
+                EndThrottled();
+            }
+            else
+            {
+                BeginThrottled();
+            }
+
+            return canExecuteActions;
+        }
+
+        public bool ActionExecutionSucceeded(bool success)
+        {
+            if (!success)
+            {
+                ActionExecutionFailed();
+            }
+            else
+            {
+                ActionExecutionSucceeded();
+            }
+
+            return success;
+        }
+
+        public bool CheckForActionCountLimitReached()
+        {
+            bool limitReached = ActionCountLimit <= ExecutionTimestamps.Count && !ActionCountSlidingWindowDuration.HasValue;
```

Comment:
While this is likely to work ok, strictly speaking `Queue<T>.Count` is not thread safe. You should still lock around it.